effekt/shared/src/main/scala/effekt/RecursiveDescent.scala

```diff
@@ -0,0 +1,1358 @@
+    val before = position
+    try { Some(p) } catch {
+      case Fail(_, _) => position = before; None
+    }
+
+  /**
+   * Tiny combinator DSL to sequence parsers
+   */
+  case class ~[+T, +U](_1: T, _2: U) {
+    override def toString = s"(${_1}~${_2})"
+  }
+
+  extension [A](self: A) {
+    @targetName("seq")
+    inline def ~[B](other: B): (A ~ B) = new ~(self, other)
+
+    @targetName("seqLeftToken")
+    inline def <~(t: TokenKind): A = { consume(t); self }
+
+    @targetName("seqLeftUnit")
+    inline def <~(t: Unit): A = { self }
+
+    inline def |(other: A): A = { backtrack(self).getOrElse(other) }
+  }
+
+  extension (self: TokenKind) {
+    @targetName("seqRightToken")
+    inline def ~>[R](other: => R): R = { consume(self); other }
+  }
+
+  extension (self: Unit) {
+    @targetName("seqRightUnit")
+    inline def ~>[R](other: => R): R = { other }
+  }
+
+  /**
+   * Repeats [[p]], separated by [[sep]] enclosed by [[before]] and [[after]]
+   */
+  inline def some[T](p: () => T, before: TokenKind, sep: TokenKind, after: TokenKind): List[T] =
+    consume(before)
+    val res = some(p, sep)
+    consume(after)
+    res
+
+  inline def some[T](p: () => T, sep: TokenKind): List[T] =
+    val components: ListBuffer[T] = ListBuffer.empty
+    components += p()
+    while (peek(sep)) {
+      consume(sep)
+      components += p()
+    }
+    components.toList
+
+  inline def someWhile[T](p: => T, lookahead: TokenKind): List[T] =
+    someWhile(p, peek(lookahead))
+
+  inline def someWhile[T](p: => T, predicate: => Boolean): List[T] =
+    val components: ListBuffer[T] = ListBuffer.empty
+    components += p
+    while (predicate) {
+      components += p
+    }
+    components.toList
+
+  inline def manyWhile[T](p: => T, lookahead: TokenKind): List[T] =
+    manyWhile(p, peek(lookahead))
+
+  inline def manyWhile[T](p: => T, predicate: => Boolean): List[T] =
+    val components: ListBuffer[T] = ListBuffer.empty
+    while (predicate) {
+      components += p
+    }
+    components.toList
+
+  inline def parens[T](p: => T): T =
+    consume(`(`)
+    val res = p
+    consume(`)`)
+    res
+
+  inline def braces[T](p: => T): T =
+    consume(`{`)
+    val res = p
+    consume(`}`)
+    res
+
+  inline def many[T](p: () => T, before: TokenKind, sep: TokenKind, after: TokenKind): List[T] =
+    consume(before)
+    if (peek(after)) {
+      consume(after)
+      Nil
+    } else {
+      val components: ListBuffer[T] = ListBuffer.empty
+      components += p()
+      while (peek(sep)) {
+        consume(sep)
+        components += p()
+      }
+      consume(after)
+      components.toList
+    }
+
+
+  // Positions
+
+  inline def nonterminal[T](inline p: => T): T = {
+    val startToken = peek
+    val start = startToken.start
+    val res = p
+    val endToken = previous
+    val end = endToken.end + 1 // since positions by lexer are inclusive, but kiama is exclusive
+
+    //    val sc: Any = res
```

Comment:
Can this be deleted?

Reply:
yes, it looks stupid but I wanted to keep it since we will definitely debug positions in the near future again ;)